README.md

```diff
@@ -145,7 +175,9 @@ code, include a script or makefile that will do the job.
 
 ## A2.5
 
-«replace this with your answer»
+The file is called sentences.cpp.
```

Comment:
I appreciate you adding code to test both positive and negative cases. However, these are actually tests, as what they do is write 1s and 0s to the console. A test will check that the code returns what is expected, and report when it doesn't.

Here's what I mean, with a fairly minimal hack to your code.

~~~ c++
using namespace std;
typedef vector<string> sentence;
typedef vector<sentence> sentences;
// ...
int main() {
	sentences valid {
	  {"the", "dog", "ate", "EOI"},
	  {"the", "lazy", "smelly", "dog", "ran", "slowly", "EOI"},
	  {"the", "lazy", "smelly", "lazy", "dog", "ate", "noisily", "EOI"},
	  {"the", "dog", "ate", "slowly", "EOI"},
	  {"the", "dog", "ran", "noisily", "EOI"},
	};
	 sentences invalid {
	  {"dog", "ate", "EOI"},
	  {"the", "lazy", "smelly", "ran", "EOI"},
	  {"the", "lazy", "smelly", "dog", "ate", "slowly", "noisily", "EOI"},
	  {"the", "dog", "EOI"},
	 };

	for (sentence s: valid) {
		for (const auto &e : s) cout << e << " "; cout << endl;
		assert(isSentence(s));
	}
	for (sentence s: invalid) {
		for (const auto &e : s) cout << e << " "; cout << endl;
		assert(!isSentence(s));
	}

	return 0;
};
~~~

# -2

Now on to the parser.

The idea of the exercise was to "Translate this table into a programming language of your choice"; that is to create an actual data structure in C++ and then traverse it during the parse. The next best alternative would be to have written the state machine totally in code, with (maybe) a function per state, or a case statement. But the code here is just a manually written parser without any mention of states. It's actually a good example of why state machines are helpful. This code has if statements nested 7 level deep. Here's a version using a table:

~~~ c++
typedef int state;
typedef std::map<std::string, state> transitions;
typedef std::map<state, transitions> state_table;

const int END = 99;

state_table states = {
  { 0, {{ "the", 1 }}},
	{ 1, {{ "lazy", 2}, {"smelly", 2}, {"dog", 3}, {"cat", 3}}},
	{ 2, {{ "lazy", 2}, {"smelly", 2}, {"dog", 3}, {"cat", 3}}},
	{ 3, {{ "ate", 4},  {"ran",    4}}},
	{ 4, {{ "slowly", 5}, { "noisily", 5}, {"EOI", END}}},
	{ 5, {{ "EOI", END }}}
};


int getNextState(int currentState, string nextWord) {
	transitions t = states[currentState];
	if (t.count(nextWord))
		return t[nextWord];
	else
		return -1;
}
~~~

# -6